File: src/native_solution/leetcode_64.rs
pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
    use std::cmp::min;
    let m = grid.len();
    let n = grid[0].len();
    let mut dp = vec![vec![0;n];m];
    for _i in 0..m{
        for _j in 0..n{
            if _i==0&&_j==0{
                dp[_i][_j] = grid[_i][_j];
            }else if _i==0 {
                dp[_i][_j] = grid[_i][_j] + dp[_i][_j-1];
            }else if _j==0{
                dp[_i][_j] = grid[_i][_j] + dp[_i-1][_j];
            }else{
                dp[_i][_j] = grid[_i][_j] + min(dp[_i][_j-1],dp[_i-1][_j]);
            }
        }
    }
    dp[m-1][n-1]
}

File: src/javascript_solution/leetcode_64.js
/**
 * 给定一个包含非负整数的 m x n 网格 grid ，请找出一条从左上角到右下角的路径，使得路径上的数字总和为最小。
 * 说明：每次只能向下或者向右移动一步。
 */

/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
  const m = grid.length;
  const n = grid[0].length;
  const dp = new Array(m).fill(0).map(() => new Array(n).fill(0));
  for (let i = 0; i < dp.length; i++) {
    for (let j = 0; j < dp[i].length; j++) {
      if (i === 0 && j === 0) {
        dp[i][j] = grid[0][0];
      } else if (i === 0) {
        dp[i][j] = dp[i][j - 1] + grid[i][j];
      } else if (j === 0) {
        dp[i][j] = dp[i - 1][j] + grid[i][j];
      } else {

        dp[i][j] = Math.min(
          dp[i - 1][j] + grid[i][j],
          dp[i][j - 1] + grid[i][j]
        );
      }
    }
  }
  return dp[m-1][n-1]
};

console.log(
  minPathSum([
    [1, 3, 1],
    [1, 5, 1],
    [4, 2, 1],
  ])
);

console.log(minPathSum([[1,2,3],[4,5,6]]))
